### PR TITLE
feat: refine swap out and migrate bucket

### PR DESCRIPTION
**Pull request by Fynnss**
### Description

refine swap out and migrate bucker

### Rationale

1. Add cancel migrate bucket , user can stop migrate bucket any time
2. Add SwapOut/CancelSwapOut/CompleteSwapOut to improve the completeness of the swapout process

### Example

NA

### Changes

Notable changes: 
* NA

### CODE REVIEW

File: x/storage/keeper/keeper.go [hunk 1]
```diff
@@ -1738,9 +1739,46 @@ func (k Keeper) garbageCollectionForResource(ctx sdk.Context, deleteStalePolicie
...
+	if !found {
+		return sptypes.ErrStorageProviderNotFound.Wrapf("src sp not found")
+	}
+
+	dstSP, found := k.spKeeper.GetStorageProvider(ctx, dstPrimarySPID)
+	if !found {
+		return sptypes.ErrStorageProviderNotFound.Wrapf("dst sp not found")
+	}
+
+	if srcSP.Status != sptypes.STATUS_IN_SERVICE || dstSP.Status != sptypes.STATUS_IN_SERVICE {
```
Comment (will-2012) [inline x/storage/keeper/keeper.go:1768]: Perfer `srcSP.IsInService()`.

File: x/storage/keeper/keeper.go [hunk 2]
```diff
@@ -1738,9 +1739,46 @@ func (k Keeper) garbageCollectionForResource(ctx sdk.Context, deleteStalePolicie
...
+		return sptypes.ErrStorageProviderNotInService.Wrapf(
+			"origin SP status: %s, dst SP status: %s", srcSP.Status.String(), dstSP.Status.String())
+	}
+
+	// check approval
+	if dstPrimarySPApproval.ExpiredHeight < (uint64)(ctx.BlockHeight()) {
```
Comment (will-2012) [inline x/storage/keeper/keeper.go:1774]: dstPrimarySPApproval.ExpiredHeight `>` (uint64)(ctx.BlockHeight()).

File: x/storage/keeper/keeper.go (hunk 1 above)
Comment (Fynnss) [reply x/storage/keeper/keeper.go]: Fixed

File: x/storage/keeper/keeper.go (hunk 2 above)
Comment (Fynnss) [reply x/storage/keeper/keeper.go]: Fixed

**Review by will-2012 — COMMENTED** [review]
LGTM.

File: x/storage/keeper/keeper.go (hunk 2 above)
Comment (will-2012) [reply x/storage/keeper/keeper.go]: Sorry, I misunderstood here; it should be `<`.

Comment (Fynnss) [reply x/storage/keeper/keeper.go]: Yes, I saw it. The e2e test case did not pass.

File: proto/greenfield/storage/events.proto [hunk 3]
```diff
@@ -466,6 +466,16 @@ message EventMigrationBucket {
   uint32 dst_primary_sp_id = 4;
 }
 
+message EventCancelMigrationBucket {
+  string operator = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
```
Comment (forcodedancing) [inline proto/greenfield/storage/events.proto:470]: Can we add a comment to explain operator here? or using `owner`?

File: proto/greenfield/storage/tx.proto [hunk 4]
```diff
@@ -558,13 +559,26 @@ message MsgMigrateBucketResponse {}
...
+  // gvg_mappings defines the src and dst gvg mapping relationships which the bucket migrate to
   repeated GVGMapping gvg_mappings = 4;
 }
 
 message MsgCompleteMigrateBucketResponse {}
+
+message MsgCancelMigrateBucket {
+  option (cosmos.msg.v1.signer) = "operator";
+  // operator defines the account address of the msg operator.
+  // Only the use can send this transaction to cancel the migrate bucket
```
Comment (forcodedancing) [inline proto/greenfield/storage/tx.proto:578]: typo `use` -> `user`

File: proto/greenfield/virtualgroup/events.proto [hunk 5]
```diff
@@ -69,6 +69,18 @@ message EventSwapOut {
...
 }
 
+message EventCompleteSwapOut {
+  uint32 storage_provider_id = 1;
+  uint32 global_virtual_group_family_id = 2;
+  repeated uint32 global_virtual_group_ids = 3;
+}
+
+message EventCancelSwapOut {
+  uint32 storage_provider_id = 1;
```
Comment (forcodedancing) [inline proto/greenfield/virtualgroup/events.proto:79]: If we want to associate to `EventSwapOut`, we need to compare all these fields. Wondering if there exists swap out `id` for this.

File: x/storage/keeper/keeper.go [hunk 6]
```diff
@@ -1738,9 +1739,46 @@ func (k Keeper) garbageCollectionForResource(ctx sdk.Context, deleteStalePolicie
...
+		return types.ErrInvalidApproval.Wrap("dst primary sp approval timeout")
+	}
+	err := k.VerifySPAndSignature(ctx, dstSP.Id, approvalBytes, dstPrimarySPApproval.Sig)
+	if err != nil {
+		return err
+	}
+
 	key := types.GetMigrationBucketKey(bucketInfo.Id)
 	if store.Has(key) {
 		panic("migration bucket key is existed.")
```
Comment (forcodedancing) [inline x/storage/keeper/keeper.go:1784]: It means the migration exists, how about return error instead, to let the user aware it?

File: x/virtualgroup/types/keys.go [hunk 7]
```diff
@@ -40,6 +40,9 @@ var (
 	GVGFamilySequencePrefix = []byte{0x33}
 
 	GVGStatisticsWithinSPKey = []byte{0x41}
+
+	SwapOutFamilyKey = []byte{0x51}
```
Comment (forcodedancing) [inline x/virtualgroup/types/keys.go:44]: Same key?

File: x/virtualgroup/keeper/keeper.go [hunk 8]
```diff
@@ -433,3 +431,110 @@ func (k Keeper) GetGlobalVirtualGroupIfAvailable(ctx sdk.Context, gvgID uint32,
...
+
+	if gvgFamilyID != types.NoSpecifiedFamilyId {
+		key := types.GetSwapOutFamilyKey(gvgFamilyID)
+		found := store.Has(key)
+		if found {
+			return types.ErrSwapOutFailed.Wrapf("SwapOutInfo of this gvg family(ID: %d) already exist", gvgFamilyID)
+
+		}
+		swapOutInfo := &types.SwapOutInfo{
+			Id:            gvgFamilyID,
```
Comment (forcodedancing) [inline x/virtualgroup/keeper/keeper.go:446]: Can we set Id to nil to save space?

File: x/virtualgroup/keeper/keeper.go [hunk 9]
```diff
@@ -433,3 +431,110 @@ func (k Keeper) GetGlobalVirtualGroupIfAvailable(ctx sdk.Context, gvgID uint32,
...
+		store.Set(key, bz)
+	} else {
+		for _, gvgID := range gvgIDs {
+			key := types.GetSwapOutGVGKey(gvgID)
+			found := store.Has(key)
+			if found {
+				return types.ErrSwapOutFailed.Wrapf("SwapOutInfo of this gvg(ID: %d) already exist.", gvgID)
+			}
+			swapOutInfo := &types.SwapOutInfo{
+				Id:            gvgID,
```
Comment (forcodedancing) [inline x/virtualgroup/keeper/keeper.go:461]: Can we set Id to nil to save space?

File: x/storage/keeper/keeper.go (hunk 6 above)
Comment (Fynnss) [reply x/storage/keeper/keeper.go:1784]: It is mentioned above to check whether the bucket is in the migrating state. So if the migration info not exist, this is the   internal error. I think its need panic here.

File: x/virtualgroup/types/keys.go (hunk 7 above)
Comment (Fynnss) [reply x/virtualgroup/types/keys.go:44]: Fixed.

File: x/virtualgroup/keeper/keeper.go (hunk 8 above)
Comment (Fynnss) [reply x/virtualgroup/keeper/keeper.go:446]: Absolutely yes.

File: x/virtualgroup/keeper/keeper.go (hunk 9 above)
Comment (Fynnss) [reply x/virtualgroup/keeper/keeper.go]: Remove the ID to save space because i think its useless.

**Review by forcodedancing — APPROVED** [review]

**Pull request merged**